core/reactor_common.c

```diff
@@ -1697,23 +1720,31 @@ lf_token_t* writable_copy(lf_token_t* token) {
     if (token->ref_count == 1) {
         DEBUG_PRINT("writable_copy: Avoided copy because reference count is %d.", token->ref_count);
         return token;
-   } else {
-        DEBUG_PRINT("writable_copy: Copying array because reference count is greater than 1. It is %d.", token->ref_count);
+    }
+    DEBUG_PRINT("writable_copy: Copying array because reference count is greater than 1. It is %d.", token->ref_count);
+    void* copy;
+    if (token->copy_constructor == NULL) {
+        DEBUG_PRINT("writable_copy: Copy constructor is NULL. Using default strategy.");
         size_t size = token->element_size * token->length;
         if (size == 0) {
             return token;
         }
-        void* copy = malloc(size);
+        copy = malloc(size);
         DEBUG_PRINT("Allocating memory for writable copy %p.", copy);
         memcpy(copy, token->value, size);
         // Count allocations to issue a warning if this is never freed.
-        _lf_count_payload_allocations++;
-        // Create a new, dynamically allocated token.
-        lf_token_t* result = create_token(token->element_size);
-        result->length = token->length;
-        result->value = copy;
-        return result;
+    } else {
+        DEBUG_PRINT("writable_copy: Copy constructor is not NULL. Using copy constructor.");
+        copy = token->copy_constructor(token->value);
```

Comment:
I wonder if we should also check here if a destructor is set and issue a warning if not.

The potential problem here is that `copy_constructor` could be using something other than `malloc` to allocate memory for `copy`. If the destructor is `NULL`, we will try to call `free` on a memory that was not allocated using `malloc`.